activesupport/lib/active_support/duration.rb

```diff
@@ -113,19 +116,49 @@ def ago(time = ::Time.current)
     def inspect #:nodoc:
       parts.
         reduce(::Hash.new(0)) { |h,(l,r)| h[l] += r; h }.
-        sort_by {|unit,  _ | [:years, :months, :days, :minutes, :seconds].index(unit)}.
+        sort_by {|unit,  _ | [:years, :months, :weeks, :days, :hours, :minutes, :seconds].index(unit)}.
         map     {|unit, val| "#{val} #{val == 1 ? unit.to_s.chop : unit.to_s}"}.
         to_sentence(locale: ::I18n.default_locale)
     end
 
+    # JSON representation should be a string formatted as ISO 8601 Duration, as recommended by Google JSON Style Guide:
+    # https://google-styleguide.googlecode.com/svn/trunk/jsoncstyleguide.xml?showone=Time_Duration_Property_Values#Time_Duration_Property_Values
     def as_json(options = nil) #:nodoc:
-      to_i
+      iso8601(precision: options && options[:precision])
     end
 
     def respond_to_missing?(method, include_private=false) #:nodoc:
       @value.respond_to?(method, include_private)
     end
 
+    # Creates a new Duration from string formatted according to ISO 8601 Duration.
+    #
+    # See http://en.wikipedia.org/wiki/ISO_8601#Durations
+    # This method allows negative parts to be present in pattern.
+    # If invalid string is provided, it will raise +ActiveSupport::Duration::ISO8601Parser::ParsingError+.
+    def self.parse!(iso8601duration)
+      parts = ISO8601Parser.new(iso8601duration).parse!
+      time  = ::Time.current
+      new(time.advance(parts) - time, parts)
```

Comment:
@pixeltrix What do you think? I can't figure out which variant is correct. For me `Time.current` is sane.

Or may be it's better not to use `Time.advance` at all but manually calculates value by doing something like this:

```ruby
parts = ActiveSupport::Duration::ISO8601Parser.new('P1Y2M3DT4H5M6S').parse!
duration = parts.map{ |unit, value| value.send(unit) } # 1.year + 2.months + 3.days + …
duration.inspect # 1 year, 2 months, 3 days, and 14706 seconds
duration.value   # 37015506.0 — different from current parse method, but stable.
duration.iso8601 # "P1Y2M3DT14706S" — hours and minutes are lost
```

But in such scenario there are lot of intermediate durations created meanwhile (as many as parts in duration, 6 in this case, not good).